src/cz/crcs/ectester/standalone/libs/jni/nettle.c

```diff
@@ -0,0 +1,512 @@
+    (*env)->ReleaseByteArrayElements(env, priv, priv_data, JNI_ABORT);
+    ecc_scalar_set(privKey, mp);
+    return priv_len;
+}
+
+JNIEXPORT jbyteArray JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeKeyAgreementSpi_00024Nettle_generateSecret___3B_3BLjava_security_spec_ECGenParameterSpec_2(JNIEnv *env, jobject self, jbyteArray pubkey, jbyteArray privkey, jobject params) {
+    jmethodID get_name = (*env)->GetMethodID(env, ecgen_parameter_spec_class, "getName", "()Ljava/lang/String;");
+    jstring name = (*env)->CallObjectMethod(env, params, get_name);
+    const char* utf_name = (*env)->GetStringUTFChars(env, name, NULL);
+    const struct ecc_curve* curve;
+    int rc;
+    char *curve_name[5] = {"secp192r1", "secp224r1", "secp256r1", "secp384r1", "secp521r1"};
+    int byte_sizes[] = {24, 28, 32, 48, 66};
+    int byte_size;
+    for (int i = 0; i < sizeof(curve_name); i++) {
+        if (strcasecmp(utf_name, curve_name[i]) == 0) {
+             curve = create_curve(env, curve_name[i]);
+             byte_size = byte_sizes[i];
+             break;
+        }
+    }
+    (*env)->ReleaseStringUTFChars(env, name, utf_name);
+    if (!curve) {
+        throw_new(env, "java/security/InvalidAlgorithmParameterException", "Curve for given bitsize not found.");
+        return NULL;
+    }
+
+    struct ecc_scalar privScalar;
+    ecc_scalar_init(&privScalar, curve);
+    barray_to_privkey(env, &privScalar, privkey);
+
+    struct ecc_point eccPubPoint;
+    ecc_point_init(&eccPubPoint, curve);
+    barray_to_pubkey(env, &eccPubPoint, pubkey);
+
+    struct ecc_point resultPoint;
+    ecc_point_init(&resultPoint, curve);
+
+    jbyteArray result = (*env)->NewByteArray(env, byte_size);
+    jbyte *result_data = (*env)->GetByteArrayElements(env, result, NULL);
+
+    native_timing_start();
+    ecc_point_mul(&resultPoint, &privScalar, &eccPubPoint);
+    native_timing_stop();
+
+    mpz_t x;
+    mpz_init(x);
+    ecc_point_get(&resultPoint, x, NULL);
+
+    size_t size;
+
+    mpz_export(NULL, &size, 1, sizeof(unsigned char), 0, 0, x);
+    int diff = byte_size - size;
+    for (int i = 0; i < diff; i++) {
+       result_data[i] = 0x00;
+    }
+    mpz_export((unsigned char*) result_data + diff, &size, 1, sizeof(unsigned char), 0, 0, x);
+    (*env)->ReleaseByteArrayElements(env, result, result_data, 0);
+    ecc_scalar_clear(&privScalar);
+    ecc_point_clear(&eccPubPoint);
+    ecc_point_clear(&resultPoint);
+    mpz_clear(x);
+    return result;
+}
+
+JNIEXPORT jobject JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeKeyAgreementSpi_00024Nettle_generateSecret___3B_3BLjava_security_spec_ECParameterSpec_2Ljava_lang_String_2(JNIEnv *env, jobject self, jbyteArray pubkey, jbyteArray privkey, jobject params, jstring algorithm) {
+    throw_new(env, "java/lang/UnsupportedOperationException", "Not supported.");
+    return NULL;
+}
+
+int signature_to_der(struct dsa_signature* signature, unsigned char *result, int byte_size) {
```

Comment:
This is also not fully correct DER signature generation and might cause issues, look at https://github.com/crocs-muni/ECTester/blob/master/src/cz/crcs/ectester/standalone/libs/jni/c_utils.c#L114 for an approach.

Reply:
Solved in 6134523